pkg/util/informermanager/informermanager_test.go

```diff
@@ -0,0 +1,792 @@
+
+		// 3. Create new FTC for daemonset
+
+		_, err := fedClient.CoreV1alpha1().FederatedTypeConfigs().Create(ctx, daemonsetFTC, metav1.CreateOptions{})
+		g.Expect(err).ToNot(gomega.HaveOccurred())
+
+		// 4. Verify that alwaysRegistered is eventually registered for the new Daemonset FTC
+
+		alwaysRegistered.ExpectGenerateEvents(daemonsetFTC.Name, 1)
+		alwaysRegistered.ExpectAddEvents(daemonsetGVK, 4)
+		alwaysRegistered.AssertEventually(g, time.Second*2)
+
+		// 5. Verify that newly generated events are also received by alwaysRegistered
+
+		generateEvents(
+			ctx,
+			g,
+			getTestDaemonSet("dm-5", "default"),
+			dynamicClient.Resource(common.DaemonSetGVR),
+			alwaysRegistered,
+		)
+
+		alwaysRegistered.AssertEventually(g, time.Second*2)
+
+		// 6. Verify that events for non-existent FTCs are not received by alwaysRegistered
+
+		generateEvents(
+			ctx,
+			g,
+			getTestSecret("sc-1", "default"),
+			dynamicClient.Resource(common.SecretGVR),
+		)
+
+		alwaysRegistered.AssertConsistently(g, time.Second*2)
+
+		// 7. Verify that unregisteredResourceEventHandler is not registered
+
+		neverRegistered.AssertConsistently(g, time.Second*2)
+	})
+
+	t.Run("EventHandlerGenerators should receive correct lastApplied and latest FTCs", func(t *testing.T) {
+		t.Parallel()
+		g := gomega.NewWithT(t)
+
+		// 1. Bootstrap environment
+
+		var generation int64 = 1
+
+		// assertionCh is used to achieve 2 things:
+		// 1. It is used to pass assertions to the main goroutine.
+		// 2. It is used as an implicit lock to ensure FTC events are not squashed by the InformerManager.
+		assertionCh := make(chan func())
+
+		ftc := deploymentFTC.DeepCopy()
+		ftc.SetGeneration(generation)
+
+		generator := &EventHandlerGenerator{
+			Predicate: func(lastApplied, latest *fedcorev1a1.FederatedTypeConfig) bool {
+				if generation == 1 {
+					assertionCh <- func() {
+						g.Expect(lastApplied).To(gomega.BeNil())
+						g.Expect(latest.GetGeneration()).To(gomega.BeNumerically("==", 1))
+					}
+				} else {
+					assertionCh <- func() {
+						g.Expect(lastApplied.GetGeneration()).To(gomega.BeNumerically("==", generation-1))
+						g.Expect(latest.GetGeneration()).To(gomega.BeNumerically("==", generation))
+					}
+				}
+
+				return true
+			},
+			Generator: func(ftc *fedcorev1a1.FederatedTypeConfig) cache.ResourceEventHandler { return nil },
+		}
+
+		defaultFTCs := []*fedcorev1a1.FederatedTypeConfig{ftc}
+		defaultObjs := []*unstructured.Unstructured{}
+		generators := []*EventHandlerGenerator{generator}
+
+		ctx, cancel := context.WithCancel(ctx)
+		manager, _, fedClient := bootstrapInformerManagerWithFakeClients(g, ctx, defaultFTCs, defaultObjs, generators)
+		defer func() {
+			cancel()
+			_ = wait.PollInfinite(time.Millisecond, func() (done bool, err error) { return manager.IsShutdown(), nil })
+		}()
+
+		fn := <-assertionCh
+		fn()
+
+		// 3. Generate FTC update events
+
+		for i := 0; i < 5; i++ {
+			generation++
```

Comment:
Is there a data race on generation?

Reply:
Line 315 is executed on an informer manager goroutine, while line 349 is executed on the main goroutine.

Reply:
Changed to use `atomic.Int64`